lib/newapi1.js

```diff
@@ -0,0 +1,319 @@
+    }
+  }
+
+  if (normalized.url === undefined) {
+    normalized.url = new URL("file:" + path.resolve(filename));
+  }
+
+  return normalized;
+}
+
+function transformOptions(options = {}, encoding) {
+  const transformed = {
+    windowOptions: {
+      // Defaults
+      url: "about:blank",
+      referrer: "",
+      contentType: "text/html",
+      parsingMode: "html",
+      userAgent: DEFAULT_USER_AGENT,
+      parseOptions: { locationInfo: false },
+      encoding,
+
+      // Defaults filled in later
+      virtualConsole: undefined,
+      cookieJar: undefined
+    },
+
+    // Defaults
+    runScripts: undefined,
+    beforeParse() { }
+  };
+
+  if (options.contentType !== undefined) {
+    const contentTypeParsed = parseContentType(options.contentType);
+    if (contentTypeParsed === null) {
+      throw new TypeError(`Could not parse the given content type of "${options.contentType}"`);
+    }
+
+    if (!contentTypeParsed.isHTML() && !contentTypeParsed.isXML()) {
+      throw new RangeError(`The given content type of "${options.contentType}" was not a HTML or XML content type`);
+    }
+
+    transformed.windowOptions.contentType = contentTypeParsed.type + "/" + contentTypeParsed.subtype;
+    transformed.windowOptions.parsingMode = contentTypeParsed.isHTML() ? "html" : "xml";
+  }
+
+  if (options.url !== undefined) {
+    transformed.windowOptions.url = (new URL(options.url)).href;
+  }
+
+  if (options.referrer !== undefined) {
+    transformed.windowOptions.referrer = (new URL(options.referrer)).href;
+  }
+
+  if (options.userAgent !== undefined) {
+    transformed.windowOptions.userAgent = String(options.userAgent);
+  }
+
+  if (options.includeNodeLocations) {
+    if (transformed.windowOptions.parsingMode === "xml") {
+      throw new TypeError("Cannot set includeNodeLocations to true with an XML content type");
+    }
+
+    transformed.windowOptions.parseOptions = { locationInfo: true };
+  }
+
+  transformed.windowOptions.cookieJar = options.cookieJar === undefined ?
+                                       new CookieJar() :
+                                       options.cookieJar;
+
+  transformed.windowOptions.virtualConsole = options.virtualConsole === undefined ?
+                                            (new VirtualConsole()).sendTo(console) :
+                                            options.virtualConsole;
+
+  if (options.runScripts !== undefined) {
+    transformed.runScripts = String(options.runScripts);
+    if (transformed.runScripts !== "dangerously" && transformed.runScripts !== "outside-only") {
+      throw new RangeError(`runScripts must be undefined, "dangerously", or "outside-only"`);
+    }
+  }
+
+  if (options.beforeParse !== undefined) {
+    transformed.beforeParse = options.beforeParse;
+  }
+
+  // concurrentNodeIterators??
+
+  return transformed;
+}
+
+function normalizeHTML(html = "") {
+  let encoding = "UTF-8";
+
+  if (ArrayBuffer.isView(html)) {
+    html = html.buffer;
+  }
```

Comment:
This doesn't work if the view has a `bufferOffset`. Check `Blob`'s [implementation](https://github.com/tmpvar/jsdom/blob/79f5522c13a2b900768183ca425ba90f90f9708d/lib/jsdom/living/file-api/Blob-impl.js#L32) for proper handling.